src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs

```diff
@@ -31,14 +31,25 @@ public HttpConnectionPools(int maxConnectionsPerServer)
         {
             _maxConnectionsPerServer = maxConnectionsPerServer;
             _pools = new ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool>();
-            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, CleanPoolTimeoutMilliseconds, CleanPoolTimeoutMilliseconds);
+            // Start out with the timer not running, since we have no pools.
+            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, Timeout.Infinite, Timeout.Infinite);
         }
 
         /// <summary>Gets a pool for the specified endpoint, adding one if none existed.</summary>
         /// <param name="key">The endpoint for the pool.</param>
         /// <returns>The retrieved pool.</returns>
-        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key) =>
-            _pools.GetOrAdd(key, (k, s) => new HttpConnectionPool(s), _maxConnectionsPerServer);
+        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key)
+        {
+            // If we currently don't have any pools, a new one will be added for this
+            // connection. We should then start reaping stale pools.
+            if (_pools.IsEmpty)
```

Comment:
> Right now we don't know if a pool was added due to the use of GetOrAdd

We don't need to be perfect here: I think it's fine for us to err on the side of running the timer versus not, which means we could just assume that if we're creating the connection (in the callback) then it's going to be added... worst case is there's a race condition, someone else adds it, and we artificially force for the timer to run.  Alternatively, yes, you could manually implement the equivalent of GetOrAdd, which is essentially just a loop that does TryGet, if it returns false does the creation and calls TryAdd, and exits the loop of the add completes.